components/brave_federated/communication_adapter.cc

```diff
@@ -0,0 +1,197 @@
+namespace brave_federated {
+
+namespace {
+
+// Maximum size of the federated server response in bytes.
+const int kMaxFederatedServerResponseSizeBytes = 1024 * 1024;  // 1 MB
+
+std::unique_ptr<network::ResourceRequest> MakeResourceRequest() {
+  auto request = std::make_unique<network::ResourceRequest>();
+  request->headers.SetHeader("Content-Type", "application/protobuf");
+  request->headers.SetHeader("Accept", "application/protobuf");
+  request->headers.SetHeader("X-Brave-FL-Federated-Learning", "?1");
+  request->credentials_mode = network::mojom::CredentialsMode::kOmit;
+  request->load_flags = net::LOAD_DO_NOT_SAVE_COOKIES;
+  request->method = net::HttpRequestHeaders::kPostMethod;
+  return request;
+}
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("federated_learning", R"(
+        semantics {
+          sender: "Federated Learning"
+          description:
+            "Federated model updates based off toy task and locally generated "
+            "synthetic dataset."
+          trigger:
+            "Reports are generated when the brave-federated flag is enabled "
+            "and learning tasks are made available by the federated server. "
+            "Reports are NOT generated when the user is on battery power "
+            "or when the device is on a metered network (not WiFi/Ethernet). "
+          data:
+            "Simple federated model updates based off toy task and locally "
+            "generated synthetic data."
+          destination: WEBSITE
+        }
+        policy {
+          cookies_allowed: NO
+          setting:
+            "This experimental feature is off by default. It can be enabled
+            "by turning on the #brave-federated experimental flag found at "
+            "'brave://flags/#brave-federated'.  "
+          policy_exception_justification:
+            "Not implemented."
+        }
+    )");
+}
+
+}  // namespace
+
+CommunicationAdapter::CommunicationAdapter(
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    const net::BackoffEntry::Policy reconnect_policy,
+    const net::BackoffEntry::Policy request_task_policy)
+    : url_loader_factory_(std::move(url_loader_factory)) {
+  reconnect_policy_ =
+      std::make_unique<const net::BackoffEntry::Policy>(reconnect_policy);
+  reconnect_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(reconnect_policy_.get());
+
+  request_task_policy_ =
+      std::make_unique<const net::BackoffEntry::Policy>(request_task_policy);
+  request_task_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(request_task_policy_.get());
+}
+
+CommunicationAdapter::~CommunicationAdapter() = default;
+
+void CommunicationAdapter::GetTasks(GetTaskCallback callback) {
+  auto request = MakeResourceRequest();
+  VLOG(2) << "FL: Requesting tasks list " << request->method << " "
+          << request->url;
+
+  const std::string payload = BuildGetTasksPayload();
```

Comment:
If `SerializeToString` can fail, then we would send an empty payload. Is this correct?

Reply:
I just checked protobuf docs and yes SerializeToString can fail, so we should be handling this case